File: exercise/java/day21/src/main/java/audit/AuditManager.java
package audit;

import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;
import java.util.List;

public class AuditManager {
    private final int maxEntriesPerFile;
    private final String directoryName;
    private final FileSystem fileSystem;

    public AuditManager(int maxEntriesPerFile, String directoryName, FileSystem fileSystem) {
        this.maxEntriesPerFile = maxEntriesPerFile;
        this.directoryName = directoryName;
        this.fileSystem = fileSystem;
    }

    public void addRecord(String visitorName, LocalDateTime timeOfVisit) {
        String[] filePaths = fileSystem.getFiles(directoryName);
        String[] sorted = sortByIndex(filePaths);
        DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        String newRecord = visitorName + ";" + timeOfVisit.format(dateTimeFormatter);

        if (sorted.length == 0) {
            String newFile = Paths.get(directoryName, "audit_1.txt").toString();
            fileSystem.writeAllText(newFile, newRecord);
            return;
        }

        int currentFileIndex = sorted.length - 1;
        String currentFilePath = sorted[currentFileIndex];
        List<String> lines = fileSystem.readAllLines(currentFilePath);

        if (lines.size() < maxEntriesPerFile) {
            lines.add(newRecord);
            String newContent = String.join(System.lineSeparator(), lines);
            fileSystem.writeAllText(currentFilePath, newContent);
        } else {
            String newName = "audit_" + (currentFileIndex + 2) + ".txt";
            String newFile = Paths.get(directoryName, newName).toString();
            fileSystem.writeAllText(newFile, newRecord);
        }
    }

    private String[] sortByIndex(String[] filePaths) {
        return Arrays.stream(filePaths)
                .sorted()
                .toArray(String[]::new);
    }
}

File: exercise/kotlin/day21/src/main/kotlin/audit/AuditManager.kt
package audit

import java.lang.System.lineSeparator
import java.nio.file.Paths
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter.ofPattern

class AuditManager(
    private val maxEntriesPerFile: Int,
    private val directoryName: String,
    private val fileSystem: FileSystem
) {
    fun addRecord(visitorName: String, timeOfVisit: LocalDateTime) {
        val filePaths = fileSystem.getFiles(directoryName)
        val sorted = sortByIndex(filePaths)
        val dateTimeFormatter = ofPattern("yyyy-MM-dd HH:mm:ss")
        val newRecord = visitorName + ";" + timeOfVisit.format(dateTimeFormatter)

        if (sorted.isEmpty()) {
            val newFile = Paths.get(directoryName, "audit_1.txt").toString()
            fileSystem.writeAllText(newFile, newRecord)
            return
        }

        val currentFileIndex = sorted.size - 1
        val currentFilePath = sorted[currentFileIndex]
        val lines: List<String> = fileSystem.readAllLines(currentFilePath)

        if (lines.size < maxEntriesPerFile) {
            val newContent = (lines + newRecord).joinToString(lineSeparator())
            fileSystem.writeAllText(currentFilePath, newContent)
        } else {
            val newName = "audit_${currentFileIndex + 2}.txt"
            val newFile = Paths.get(directoryName, newName).toString()
            fileSystem.writeAllText(newFile, newRecord)
        }
    }

    private fun sortByIndex(filePaths: List<String>): List<String> = filePaths.sorted()
}
